File: Sources/CritterbarCore/Critter.swift
import Foundation

public enum CritterType: String, CaseIterable {
    case cat
    case dog
    case bird

    public var emoji: String {
        switch self {
        case .cat: return "🐱"
        case .dog: return "🐶"
        case .bird: return "🐦"
        }
    }

    public var baseSpeed: Double {
        switch self {
        case .cat: return 25.0
        case .dog: return 35.0
        case .bird: return 45.0
        }
    }

    public var sniffEmoji: String {
        switch self {
        case .cat: return "😺"
        case .dog: return "🐕"
        case .bird: return "🐤"
        }
    }
}

public struct CritterBounds {
    public let minX: Double
    public let minY: Double
    public let maxX: Double
    public let maxY: Double

    public var width: Double { maxX - minX }
    public var height: Double { maxY - minY }

    public init(minX: Double, minY: Double, maxX: Double, maxY: Double) {
        self.minX = minX
        self.minY = minY
        self.maxX = maxX
        self.maxY = maxY
    }
}

public enum Edge: Int, CaseIterable {
    case bottom = 0
    case right = 1
    case top = 2
    case left = 3
}

public enum CritterState {
    case walking
    case sniffing(remaining: Double)
}

public class Critter: Identifiable {
    public let id = UUID()
    public let type: CritterType

    public var x: Double
    public var y: Double
    public var screenIndex: Int

    public private(set) var edge: Edge
    public private(set) var movingForward: Bool // true = clockwise along edges
    public private(set) var state: CritterState = .walking
    public var speed: Double

    private var sniffTimer: Double = 0
    private var nextSniffIn: Double

    public init(type: CritterType, x: Double, y: Double, screenIndex: Int = 0) {
        self.type = type
        self.x = x
        self.y = y
        self.screenIndex = screenIndex
        self.speed = type.baseSpeed
        self.edge = Edge.allCases.randomElement()!
        self.movingForward = Bool.random()
        self.nextSniffIn = Double.random(in: 3.0...8.0)
    }

    public var displayEmoji: String {
        return type.emoji
    }

    public func snapToEdge(bounds: CritterBounds) {
        let margin: Double = 0
        switch edge {
        case .bottom:
            y = bounds.minY + margin
            x = x.clamped(to: bounds.minX...bounds.maxX)
        case .top:
            y = bounds.maxY - margin - 48
            x = x.clamped(to: bounds.minX...bounds.maxX)
        case .left:
            x = bounds.minX + margin
            y = y.clamped(to: bounds.minY...bounds.maxY)
        case .right:
            x = bounds.maxX - margin - 48
            y = y.clamped(to: bounds.minY...bounds.maxY)
        }
    }

    public func update(deltaTime: Double, bounds: CritterBounds) {
        switch state {
        case .sniffing(let remaining):
            let newRemaining = remaining - deltaTime
            if newRemaining <= 0 {
                state = .walking
                nextSniffIn = Double.random(in: 3.0...8.0)
                sniffTimer = 0
            } else {
                state = .sniffing(remaining: newRemaining)
            }
            return

        case .walking:
            sniffTimer += deltaTime
            if sniffTimer >= nextSniffIn {
                state = .sniffing(remaining: Double.random(in: 1.0...3.0))
                sniffTimer = 0
                return
            }
        }

        let movement = speed * deltaTime
        let direction: Double = movingForward ? 1.0 : -1.0

        // Nudge inward after corner transitions to prevent oscillation
        let cornerNudge: Double = 2.0

        switch edge {
        case .bottom:
            x += movement * direction
            y = bounds.minY
            if x >= bounds.maxX - 48 {
                x = bounds.maxX - 48
                y = bounds.minY + cornerNudge
                edge = movingForward ? .right : .left
            } else if x <= bounds.minX {
                x = bounds.minX
                y = bounds.minY + cornerNudge
                edge = movingForward ? .left : .right
            }

        case .right:
            y += movement * direction
            x = bounds.maxX - 48
            if y >= bounds.maxY - 48 {
                y = bounds.maxY - 48
                x = bounds.maxX - 48 - cornerNudge
                edge = movingForward ? .top : .bottom
            } else if y <= bounds.minY {
                y = bounds.minY
                x = bounds.maxX - 48 - cornerNudge
                edge = movingForward ? .bottom : .top
            }

        case .top:
            x -= movement * direction
            y = bounds.maxY - 48
            if x <= bounds.minX {
                x = bounds.minX
                y = bounds.maxY - 48 - cornerNudge
                edge = movingForward ? .left : .right
            } else if x >= bounds.maxX - 48 {
                x = bounds.maxX - 48
                y = bounds.maxY - 48 - cornerNudge
                edge = movingForward ? .right : .left
            }

        case .left:
            y -= movement * direction
            x = bounds.minX
            if y <= bounds.minY {
                y = bounds.minY
                x = bounds.minX + cornerNudge
                edge = movingForward ? .bottom : .top
            } else if y >= bounds.maxY - 48 {
                y = bounds.maxY - 48
                x = bounds.minX + cornerNudge
                edge = movingForward ? .top : .bottom
            }
        }
    }
}

extension Double {
    func clamped(to range: ClosedRange<Double>) -> Double {
        return Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
    }
}

File: src/critter.ts
export type CritterTypeName = "cat" | "dog" | "bird" | "rabbit" | "hamster" | "fox" | "frog" | "turtle";

const NAME_POOLS: Record<CritterTypeName, string[]> = {
  cat:     ["Captain Whiskers", "Purrlock Holmes", "Whisker the Wizard", "Meowgician", "Sir Pounce-a-Lot", "Meowrio", "Clawdia", "Cat-a-pult"],
  dog:     ["Sir Barksalot", "Indiana Bones", "Chewbarka", "Pawadin", "Bark the Barbarian", "Dogmeat", "K-9", "Woofgang"],
  bird:    ["Chirpy McChirpface", "Hawkward the Ranger", "Talon Darkbeak", "Flappy Bird", "Kazooie", "Falco", "Nest.js", "Robin Hood"],
  rabbit:  ["Hop Solo", "Bunnomancer", "Hare of Holding", "Thumper the Rogue", "Critical Hop", "/bin/bunny", "Thumper McFluff", "Bunzilla"],
  hamster: ["Boo", "Hamster of Vecna", "Squeakthief", "Sir Squeaks", "Nibbles", "sudo squeak", "Hamtaro", "Wheel of Fortune"],
  fox:     ["Firefox", "Sly Cooper", "Reynard the Trickster", "Foxblade", "Star Fox", "Tails", "Fox Mulder", "Vulpix the Sorcerer"],
  frog:    ["Frogger", "Battletoads", "Frogmire the Warlock", "Grung Overlord", "Sir Hopsalot", "Toadally Awesome", "Kermit the Vibes", "Chrono Toad"],
  turtle:  ["Tortle Monk", "Shellazar the Wise", "Bowser", "Koopa Troopa", "Shelly McShellface", "Tortellini", "Turbo the Slow", "Shell Script"],
};

export function generateName(type: CritterTypeName): string {
  const pool = NAME_POOLS[type];
  return pool[Math.floor(Math.random() * pool.length)];
}

export const CRITTER_TYPES: Record<
  CritterTypeName,
  { emoji: string; speed: number }
> = {
  turtle:  { emoji: "🐢", speed: 5 },
  hamster: { emoji: "🐹", speed: 9 },
  cat:     { emoji: "🐱", speed: 12 },
  frog:    { emoji: "🐸", speed: 14 },
  dog:     { emoji: "🐶", speed: 17 },
  rabbit:  { emoji: "🐰", speed: 20 },
  bird:    { emoji: "🐦", speed: 22 },
  fox:     { emoji: "🦊", speed: 30 },
};

export const CRITTER_SIZE = 24;

export interface CritterBounds {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

export type Edge = "bottom" | "right" | "top" | "left";

export type CritterState =
  | { kind: "walking" }
  | { kind: "sniffing"; remaining: number };

let nextId = 0;

// Speed multiplier options: weighted towards normal (1.0)
const SPEED_MULTIPLIERS = [0.5, 0.75, 1.0, 1.0, 1.0, 1.25, 1.5];

export class Critter {
  readonly id: number;
  readonly type: CritterTypeName;
  readonly emoji: string;
  readonly baseSpeed: number;
  readonly name: string;

  x: number;
  y: number;
  edge: Edge;
  movingForward: boolean;
  state: CritterState = { kind: "walking" };
  speed: number;

  private sniffTimer = 0;
  private nextSniffIn: number;
  private speedTimer = 0;
  private nextSpeedChangeIn: number;

  constructor(type: CritterTypeName, x: number, y: number, name?: string) {
    this.id = nextId++;
    this.type = type;
    this.emoji = CRITTER_TYPES[type].emoji;
    this.baseSpeed = CRITTER_TYPES[type].speed;
    this.speed = this.baseSpeed;
    this.name = name ?? generateName(type);
    this.x = x;
    this.y = y;

    const edges: Edge[] = ["bottom", "right", "top", "left"];
    this.edge = edges[Math.floor(Math.random() * edges.length)];
    this.movingForward = Math.random() < 0.5;
    this.nextSniffIn = 3 + Math.random() * 5;
    this.nextSpeedChangeIn = 5 + Math.random() * 10;
  }

  snapToEdge(bounds: CritterBounds): void {
    switch (this.edge) {
      case "bottom":
        this.y = bounds.minY;
        this.x = clamp(this.x, bounds.minX, bounds.maxX);
        break;
      case "top":
        this.y = bounds.maxY - CRITTER_SIZE;
        this.x = clamp(this.x, bounds.minX, bounds.maxX);
        break;
      case "left":
        this.x = bounds.minX;
        this.y = clamp(this.y, bounds.minY, bounds.maxY);
        break;
      case "right":
        this.x = bounds.maxX - CRITTER_SIZE;
        this.y = clamp(this.y, bounds.minY, bounds.maxY);
        break;
    }
  }

  update(deltaTime: number, bounds: CritterBounds): void {
    // Speed variation runs regardless of sniff state
    this.speedTimer += deltaTime;
    if (this.speedTimer >= this.nextSpeedChangeIn) {
      this.speedTimer = 0;
      this.nextSpeedChangeIn = 5 + Math.random() * 10;
      const multiplier = SPEED_MULTIPLIERS[Math.floor(Math.random() * SPEED_MULTIPLIERS.length)];
      this.speed = this.baseSpeed * multiplier;
    }

    if (this.state.kind === "sniffing") {
      const remaining = this.state.remaining - deltaTime;
      if (remaining <= 0) {
        this.state = { kind: "walking" };
        this.nextSniffIn = 3 + Math.random() * 5;
        this.sniffTimer = 0;
      } else {
        this.state = { kind: "sniffing", remaining };
      }
      return;
    }

    this.sniffTimer += deltaTime;
    if (this.sniffTimer >= this.nextSniffIn) {
      this.state = { kind: "sniffing", remaining: 2 + Math.random() * 3 };
      this.sniffTimer = 0;
      return;
    }

    const movement = this.speed * deltaTime;
    const dir = this.movingForward ? 1 : -1;

    // Corner transitions: the side you hit determines the next edge.
    // movingForward only affects travel direction, not which edge to go to.
    switch (this.edge) {
      case "bottom":
        this.x += movement * dir;
        this.y = bounds.minY;
        if (this.x >= bounds.maxX - CRITTER_SIZE) {
          this.x = bounds.maxX - CRITTER_SIZE;
          this.edge = "right";
        } else if (this.x <= bounds.minX) {
          this.x = bounds.minX;
          this.edge = "left";
        }
        break;

      case "right":
        this.y += movement * dir;
        this.x = bounds.maxX - CRITTER_SIZE;
        if (this.y >= bounds.maxY - CRITTER_SIZE) {
          this.y = bounds.maxY - CRITTER_SIZE;
          this.edge = "top";
        } else if (this.y <= bounds.minY) {
          this.y = bounds.minY;
          this.edge = "bottom";
        }
        break;

      case "top":
        this.x -= movement * dir;
        this.y = bounds.maxY - CRITTER_SIZE;
        if (this.x <= bounds.minX) {
          this.x = bounds.minX;
          this.edge = "left";
        } else if (this.x >= bounds.maxX - CRITTER_SIZE) {
          this.x = bounds.maxX - CRITTER_SIZE;
          this.edge = "right";
        }
        break;

      case "left":
        this.y -= movement * dir;
        this.x = bounds.minX;
        if (this.y <= bounds.minY) {
          this.y = bounds.minY;
          this.edge = "bottom";
        } else if (this.y >= bounds.maxY - CRITTER_SIZE) {
          this.y = bounds.maxY - CRITTER_SIZE;
          this.edge = "top";
        }
        break;
    }

    // Always clamp to bounds regardless of cooldown
    this.x = clamp(this.x, bounds.minX, bounds.maxX - CRITTER_SIZE);
    this.y = clamp(this.y, bounds.minY, bounds.maxY - CRITTER_SIZE);
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}
